src/cmd/metadata/mod.rs

```diff
@@ -0,0 +1,367 @@
+// Copyright 2018-2020 Parity Technologies (UK) Ltd.
+// This file is part of cargo-contract.
+//
+// cargo-contract is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// cargo-contract is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with cargo-contract.  If not, see <http://www.gnu.org/licenses/>.
+
+mod contract;
+
+use crate::{
+    crate_metadata::CrateMetadata,
+    util,
+    workspace::{ManifestPath, Workspace},
+    UnstableFlags, Verbosity,
+};
+use anyhow::Result;
+use contract::{
+    Compiler, Contract, ContractMetadata, Language, Source, SourceCompiler, SourceLanguage, User,
+};
+use semver::Version;
+use std::{fs, path::PathBuf};
+use url::Url;
+
+const METADATA_FILE: &str = "metadata.json";
+
+/// Executes the metadata generation process
+struct GenerateMetadataCommand {
+    crate_metadata: CrateMetadata,
+    verbosity: Option<Verbosity>,
+    unstable_options: UnstableFlags,
+}
+
+impl GenerateMetadataCommand {
+    pub fn exec(&self) -> Result<PathBuf> {
+        util::assert_channel()?;
+        println!("  Generating metadata");
+
+        let cargo_meta = &self.crate_metadata.cargo_meta;
+        let out_path = cargo_meta.target_directory.join(METADATA_FILE);
+        let target_dir = cargo_meta.target_directory.clone();
+
+        // build the extended contract project metadata
+        let (source_meta, contract_meta, user_meta) = self.extended_metadata()?;
+
+        let generate_metadata = |manifest_path: &ManifestPath| -> Result<()> {
+            let target_dir_arg = format!("--target-dir={}", target_dir.to_string_lossy());
+            let stdout = util::invoke_cargo(
+                "run",
+                &[
+                    "--package",
+                    "metadata-gen",
+                    &manifest_path.cargo_arg(),
+                    &target_dir_arg,
+                    "--release",
+                    // "--no-default-features", // Breaks builds for MacOS (linker errors), we should investigate this issue asap!
```

Comment:
Have we investigated or turned this into an issue, yet?

Reply:
But also I think it is fine that we do not compile the `metadata-gen` with `--no-default-features`. Why did we in the first place?

Reply:
I've no idea this comment is leftover from Sergei's initial work. I can remove this since it does not seem necessary (we now control the generation of the `metadata-gen` package after all)